### ISSUE TITLE
Emacs: error while loading shared libraries: ?: cannot open shared object file: No such file or directory

### ISSUE THREAD

**Issue by rvasin** [msys2/MSYS2-packages#2173]
After installing updates on October 2 I started to have this error when running emacs:

C:/msys32/usr/bin/emacs.exe: error while loading shared libraries: ?: cannot open shared object file: No such file or directory

Emacs is installed as msys package:
msys/emacs 26.3-1 (editors) [installed]

The same issued #1622 was reported as fixed in April 2019. Does it mean the someone from maintainers broke the dependences again?

**Comment by lazka** [comment]
We no longer support 32bit msys since may 2020. But Alexey update some packages some days ago, so maybe he missed a rebuild.

https://www.msys2.org/news/#2020-05-17-32-bit-msys2-no-longer-actively-supported

**Comment by rvasin** [comment]
OK. Thank you for noticing about the drop of support 32bit MSYS. I missed this news announce. I am on 64bit Windows and I  installed MSYS2 maybe 8 years ago. Everything worked fine until I installed updates on Oct 2 (and did regular "one a week" updates until that date). 

Emacs is located in 3 packages:
mingw32/mingw-w64-i686-emacs 27.1-1
mingw64/mingw-w64-x86_64-emacs 27.1-1
msys/emacs 26.3-1 (editors)

So i will probably try mingw's version of Emacs after the issue  #2171 with repository availability will be fixed.

**Comment by jeremyd2019** [comment]
I opened #2067 for the drop of 32-bit support.  Pretty much for my own use, I have been building msys package updates for 32-bit and maintaining a repo of those updated packages.  https://github.com/msys2/MSYS2-packages/issues/2067#issuecomment-668214918

**Comment by dscho** [comment]
@rvasin just for shiggles, could you run `ldd /usr/bin/emacs.exe` and see whether it lists either hogweed and/or nettle as a dependency?

**Comment by rvasin** [comment]
@dscho: ldd /usr/bin/emacs.exe
output is:
 ntdll.dll => /c/Windows/SysWOW64/ntdll.dll (0x76ee0000)
        kernel32.dll => /c/Windows/syswow64/kernel32.dll (0x74bc0000)
        KERNELBASE.dll => /c/Windows/syswow64/KERNELBASE.dll (0x74b70000)
        msys-gio-2.0-0.dll => /usr/bin/msys-gio-2.0-0.dll (0x6acc0000)
        msys-gmodule-2.0-0.dll => /usr/bin/msys-gmodule-2.0-0.dll (0x61d80000)
        msys-glib-2.0-0.dll => /usr/bin/msys-glib-2.0-0.dll (0x6a200000)
        msys-iconv-2.dll => /usr/bin/msys-iconv-2.dll (0x618c0000)
        msys-2.0.dll => /usr/bin/msys-2.0.dll (0x61000000)
        msys-intl-8.dll => /usr/bin/msys-intl-8.dll (0x6d0f0000)
        msys-gcc_s-1.dll => /usr/bin/msys-gcc_s-1.dll (0x6ac00000)
        msys-pcre-1.dll => /usr/bin/msys-pcre-1.dll (0x64bc0000)
        msys-gobject-2.0-0.dll => /usr/bin/msys-gobject-2.0-0.dll (0x6e200000)
        msys-ffi-7.dll => /usr/bin/msys-ffi-7.dll (0x69000000)
        msys-z.dll => /usr/bin/msys-z.dll (0x6c520000)
        msys-gnutls-30.dll => /usr/bin/msys-gnutls-30.dll (0x61e40000)
        msys-gmp-10.dll => /usr/bin/msys-gmp-10.dll (0x63d80000)

**Comment by dscho** [comment]
Right, and `msys-gnutls-30.dll` is broken by a recent `nettle` upgrade... Hopefully we'll get `gnutls` rebuilt (we're [discussing this on Gitter](https://gitter.im/msys2/msys2?at=5f873e34a1c81d0a7ee1457f)). /cc @elieux

**Comment by rvasin** [comment]
BTW the same  error  is displayed when I am trying to run mutt (email client). 
The output of ldd:

$ ldd /usr/bin/mutt
        ntdll.dll => /c/Windows/SysWOW64/ntdll.dll (0x76ee0000)
        kernel32.dll => /c/Windows/syswow64/kernel32.dll (0x74bc0000)
        KERNELBASE.dll => /c/Windows/syswow64/KERNELBASE.dll (0x74b70000)
        msys-gcc_s-1.dll => /usr/bin/msys-gcc_s-1.dll (0x6ac00000)
        msys-2.0.dll => /usr/bin/msys-2.0.dll (0x61000000)
        msys-gdbm-6.dll => /usr/bin/msys-gdbm-6.dll (0x68fc0000)
        msys-intl-8.dll => /usr/bin/msys-intl-8.dll (0x6d0f0000)
        msys-iconv-2.dll => /usr/bin/msys-iconv-2.dll (0x618c0000)
        msys-gnutls-30.dll => /usr/bin/msys-gnutls-30.dll (0x61e40000)
        msys-gmp-10.dll => /usr/bin/msys-gmp-10.dll (0x63d80000)

**Comment by dscho** [comment]
@rvasin you can work around that by getting the old dependencies somehow (`msys-hogweed-5.dll` and `msys-nettle-7.dll` e.g. by downloading them from https://github.com/git-for-windows/git-sdk-32/blob/main/usr/bin/msys-hogweed-5.dll and https://github.com/git-for-windows/git-sdk-32/blob/main/usr/bin/msys-nettle-7.dll).

**Comment by dscho** [comment]
@rvasin if you'd prefer to install packages, Git for Windows' version of `libgnutls` _might_ fix this for you: https://wingit.blob.core.windows.net/i686/libgnutls-3.6.15-1-i686.pkg.tar.xz (there is also a `.sig` file if you want my signature).

**Comment by abcfy2** [comment]
Hi there.

I'm facing the same issue with `curl` package now:

```sh
$ /usr/bin/curl --version
C:/msys64/usr/bin/curl.exe: error while loading shared libraries: ?: cannot open shared object file: No such file or directory
```

```sh
$ ldd /usr/bin/curl.exe
        ntdll.dll => /c/WINDOWS/SYSTEM32/ntdll.dll (0x7ffcf0260000)
        KERNEL32.DLL => /c/WINDOWS/System32/KERNEL32.DLL (0x7ffceeec0000)
        KERNELBASE.dll => /c/WINDOWS/System32/KERNELBASE.dll (0x7ffcedad0000)
        msys-curl-4.dll => /usr/bin/msys-curl-4.dll (0x3ff7d0000)
        msys-2.0.dll => /usr/bin/msys-2.0.dll (0x210040000)
        msys-z.dll => /usr/bin/msys-z.dll (0x3fa990000)
        msys-crypto-1.1.dll => /usr/bin/msys-crypto-1.1.dll (0x3ff880000)
        msys-gssapi-3.dll => /usr/bin/msys-gssapi-3.dll (0x3fe5e0000)
        msys-idn2-0.dll => /usr/bin/msys-idn2-0.dll (0x3fc080000)
        msys-nghttp2-14.dll => /usr/bin/msys-nghttp2-14.dll (0x3fbc40000)
        msys-psl-5.dll => /usr/bin/msys-psl-5.dll (0x3fb5c0000)
        msys-ssh2-1.dll => /usr/bin/msys-ssh2-1.dll (0x3fb030000)
        msys-ssl-1.1.dll => /usr/bin/msys-ssl-1.1.dll (0x3fafa0000)
        msys-asn1-8.dll => /usr/bin/msys-asn1-8.dll (0x3ffcf0000)
        msys-com_err-1.dll => /usr/bin/msys-com_err-1.dll (0x3ffb10000)
        msys-heimntlm-0.dll => /usr/bin/msys-heimntlm-0.dll (0x3fe530000)
        msys-heimbase-1.dll => /usr/bin/msys-heimbase-1.dll (0x3fe540000)
        msys-iconv-2.dll => /usr/bin/msys-iconv-2.dll (0x3fe370000)
        msys-hcrypto-4.dll => /usr/bin/msys-hcrypto-4.dll (0x3fe590000)
        msys-krb5-26.dll => /usr/bin/msys-krb5-26.dll (0x3fbf70000)
        msys-intl-8.dll => /usr/bin/msys-intl-8.dll (0x3fc060000)
        msys-roken-18.dll => /usr/bin/msys-roken-18.dll (0x3fb200000)
        msys-unistring-5.dll => /usr/bin/msys-unistring-5.dll (0x3fabb0000)
        msys-unistring-5.dll => /usr/bin/msys-unistring-5.dll (0xea0000)
        msys-wind-0.dll => /usr/bin/msys-wind-0.dll (0x3fab60000)
        msys-gcc_s-seh-1.dll => /usr/bin/msys-gcc_s-seh-1.dll (0x3ff310000)
        msys-wind-0.dll => /usr/bin/msys-wind-0.dll (0x640000)
        msys-sqlite3-0.dll => /usr/bin/msys-sqlite3-0.dll (0x3fb070000)
        msys-crypt-0.dll => /usr/bin/msys-crypt-0.dll (0x3ffaf0000)
        msys-hx509-5.dll => /usr/bin/msys-hx509-5.dll (0x3fe480000)
```

How could I solve this issue ? Thanks.

**Comment by Biswa96** [comment]
@abcfy2 Does updating packages with `pacman -Syyu` command resolves the issue?

**Comment by abcfy2** [comment]
> @abcfy2 Does updating packages with `pacman -Syyu` command resolves the issue?

Thanks. I got this error:

```sh
$ LC_ALL=en_US.UTF-8 pacman -Syyu
error: mingw32: signature from "Christoph Reiter (MSYS2 development key) <reiter.christoph@gmail.com>" is invalid
:: Synchronizing package databases...
 git-for-windows                              52.8 KiB  44.7 KiB/s 00:01 [#######################################] 100%
 mingw32                                    1734.9 KiB  2.64 MiB/s 00:01 [#######################################] 100%
 mingw64                                    1791.7 KiB  2.77 MiB/s 00:01 [#######################################] 100%
 ucrt64                                     1836.7 KiB  2.31 MiB/s 00:01 [#######################################] 100%
 clang64                                    1781.0 KiB  2.64 MiB/s 00:01 [#######################################] 100%
 msys                                        419.1 KiB  3.56 MiB/s 00:00 [#######################################] 100%
:: Starting core system upgrade...
 there is nothing to do
:: Starting full system upgrade...
warning: vim: local (9.0.1182-1) is newer than msys (8.2.5117-2)
 there is nothing to do
```

**Comment by Biswa96** [comment]
It seems that you are using git-for-windows. If yes, follow this documentation https://github.com/git-for-windows/git/wiki/Install-inside-MSYS2-proper

**Comment by abcfy2** [comment]
> It seems that you are using git-for-windows. If yes, follow this documentation https://github.com/git-for-windows/git/wiki/Install-inside-MSYS2-proper

Thanks. I try to remove `git-for-windows` repo and set `SigLevel    = Never` in `/etc/pacman.conf`. Then use

```sh
pacman -Syyuu
```

After install amount of packages, but I still get this error:

```sh
$ /usr/bin/curl.exe --version
C:/msys64/usr/bin/curl.exe: error while loading shared libraries: ?: cannot open shared object file: No such file or directory

```

**Comment by Biswa96** [comment]
I am not sure what exactly is happening there. @dscho, one of the maintainer of git-for-windows, may help.

**Comment by lazka** [comment]
As it is not supported, this also means you wont get any support.

I'll try to reproduce in a fresh msys2 install though, to rule that out.

**Comment by dscho** [comment]
Indeed, this is one of the pitfalls of that unsupported route, and of not even considering to help by stepping up and trying to offer assistance in supporting it but instead leaning on others to do that for you.

In this instance, `ldd`ing the cURL _DLL_ instead of the `.exe` will show you the problem: Git for Windows' OpenSSL is still at v1.1.* while MSYS2's has advanced to v3.*. This version is encoded in the DLL name. So if you still have `libcurl` from Git for Windows (`pacman -Qi` shows the packager if you want to check), but `libssl` from MSYS2, things are broken in the setup.

**Comment by abcfy2** [comment]
> Indeed, this is one of the pitfalls of that unsupported route, and of not even considering to help by stepping up and trying to offer assistance in supporting it but instead leaning on others to do that for you.
> 
> In this instance, `ldd`ing the cURL _DLL_ instead of the `.exe` will show you the problem: Git for Windows' OpenSSL is still at v1.1.* while MSYS2's has advanced to v3.*. This version is encoded in the DLL name. So if you still have `libcurl` from Git for Windows (`pacman -Qi` shows the packager if you want to check), but `libssl` from MSYS2, things are broken in the setup.

Thanks. Is there any way to fix `curl`? I've removed `git-for-windows` repo and use `pacman -Syyuu` to force sync with official repo, but I still got this issue.

**Comment by Biswa96** [comment]
I think the proper solution would be to install and use msys2 and git-for-windows separately. Things will be in their own installation path. Though there would be some duplication.
